### ISSUE TITLE
Suggest better repeated macro param declaration pattern

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#9959]
### What it does

Try to catch bad macro param patterns, and suggest better ones:
* Replace `<anything>, $($e:expr),*` with `<anything> $(, $e:expr)* $(,)?`
  * this must also replace all usages of `<anything>, $($e,)*` with `<anything> $(, $e)*`
* Add missing trailing comma - can be fixed by adding `$(,)?`.  See also #1848
* ? some other bad patterns ?

### Lint Name

bad_macro_param_pattern

### Category

suspicious

### Advantage

- keeps macro declarations more consistent
- catches common errors like using format-like macro: `my_macro!("{}", foo)` not being replaceable by `my_macro!("{foo}")` (i.e. it is no longer requires a single comma to always be present)

### Drawbacks

possible formatting issues on auto-fix

### Example

```rust
macro_rules! my_macro {
    ($fmt:literal, $($e:expr),*) => {
        println!($fmt, $($e,)*)
    }
}
```

Could be written as:

```rust
macro_rules! my_macro {
    ($fmt:literal $(, $e:expr)* $(,)?) => {
        println!($fmt $(, $e)*)
    }
}
```

**Comment by mi2ebi** [comment]
i admit i don't see why this is better

**Comment by nyurik** [comment]
Allows optional trailing comma

**Comment by mi2ebi** [comment]
well yes but why replace `, $($e:expr),*` with `$(, $e:expr)*`?
